endpoints/openrtb2/video_auction.go

```diff
@@ -142,6 +142,29 @@ func (deps *endpointDeps) VideoAuctionEndpoint(w http.ResponseWriter, r *http.Re
 			errL = []error{err}
 			return
 		}
+		if bidReq.App == nil && bidReq.Site == nil {
+			err = fmt.Errorf("Invalid JSON in Default Request Settings: %s", err)
+			errL = []error{err}
+			return
+		} else if bidReq.App.Publisher == nil && bidReq.Site.Publisher == nil {
+			err = fmt.Errorf("Invalid JSON in Default Request Settings: %s", err)
+			errL = []error{err}
+			return
+		} else if bidReq.App != nil && bidReq.Site != nil {
+			err = fmt.Errorf("Request specifies both 'Site' and 'App' values", err)
+			errL = []error{err}
+			return
+		} else if bidReq.App != nil {
+			labels.Source = pbsmetrics.DemandApp
+			if bidReq.App.Publisher != nil && bidReq.App.Publisher.ID != "" {
+				labels.PubID = bidReq.App.Publisher.ID
+			}
+		} else { // bidReq.Site != nil
+			labels.Source = pbsmetrics.DemandWeb
+			if bidReq.Site.Publisher != nil && bidReq.Site.Publisher.ID != "" {
+				labels.PubID = bidReq.Site.Publisher.ID
+			}
+		}
```

Comment:
1) this code block shouldn't be run until after videoBidReq is merged into bidReq; note that labels.Source and labels.PubID will never be populated if no deps.default.Request
2) this code is very inefficient:
- three tests for nil-ness of bidReq.App will be run in every non-error case
- two tests for nil-ness of bidReq.Site will be run in every non-error case
3) line 146:  err passed as arg to fmt.Errorf has no value, wrong error message
4) line 149:  will always panic for correct input given that at least one of bidReq.App and bidReq.Site must be nil
5) line 149:  not an error if neither Publisher value is set
6) line 150:  err passed as arg to fmt.Errorf has no value, though not an error
7) line 154:  err passed as arg to fmt.Errorf has no value

This is roughly how I would recommend writing the above block of code:

if bidReq.App != nil {
        if bidReq.Site != nil {
                // set error1 here -- both are set, probably via errors.New()
                return
        }
        labels.Source = pbsmetrics.DemandApp
        if bidReq.App.Publisher != nil && bidReq.App.Publisher.ID != "" {
                labels.PubID = bidReq.App.Publisher.ID
        }
} else {
        if bidReq.Site == nil {
                // set error2 here -- neither is set, probably via errors.New()
                return
        }
        labels.Source = pbsmetrics.DemandWeb
        if bidReq.Site.Publisher != nil && bidReq.Site.Publisher.ID != "" {
                labels.PubID = bidReq.Site.Publisher.ID
        }
}

Reply:
1) three tests for nil-ness of bidReq.App will be run in every non-error case
       Lines 145 to 152 were left as is. Their modification is out of the scope of this PR and I personally don't mind the current if/else layout
       because it delimits every scenario in a more readable way.
    2) two tests for nil-ness of bidReq.Site will be run in every non-error case
       Same as above.
    3) line 146: err passed as arg to fmt.Errorf has no value, wrong error message
       I didn't write this part of the code, correcting those errors should be handled in a separate issue unless otherwise instructed
    4) line 149: will always panic for correct input given that at least one of bidReq.App and bidReq.Site must be nil
       Line was already sitting there. I added lines 153 to 167.
    5) line 149: not an error if neither Publisher value is set
       Line was already sitting there. Since this part of the code has been code reviewed, approved and in production already, I
       assumed it to be correct.
    6) line 150: err passed as arg to fmt.Errorf has no value, though not an error
       Doesn't affect anything though
    7) line 154: err passed as arg to fmt.Errorf has no value
       You are right, forgot to add the '%s'. No errors were thrown though.
    8) This is roughly how I would recommend writing the above block of code:
       So, it's basically the same approach I used from line 157 on, but I didn't do it because it omits the check in line 149 which, given this
       code is sitting in production and has been previously reviewed and code approved, I didn't touch because I assumed it behaved as desired

Reply:
Joe I confused this PR with another one. Went back to a previous approach that corrects the issues you listed. I'll wait for any feedback